### PR TITLE
Wallet refactoring leading up to multiwallet

### PR DESCRIPTION
**Pull request by luke-jr**
Part of the refactorings needed for basic multiwallet (#8694)

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3457,7 +3471,18 @@ bool CWallet::InitLoadWallet()
...
 
-    pwalletMain = walletInstance;
+    return walletInstance;
+}
+
+bool CWallet::InitLoadWallet()
+{
+    std::string walletFile = GetArg("-wallet", DEFAULT_WALLET_DAT);
+
+    CWallet * const pwallet = CreateWalletFromFile(walletFile);
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: nit: `const CWallet *pwallet = `

**Review by jonasschnelli — APPROVED** [review]
Code Review ACK https://github.com/bitcoin/bitcoin/pull/8776/commits/682936cd8f2184917aea70174fb8ac5e9cca85f6

**Comment by jonasschnelli** [comment]
Needs rebase.

**Comment by luke-jr** [comment]
Rebased

**Comment by MarcoFalke** [comment]
utACK 24539dc21fe5969c619f20

**Comment by MarcoFalke** [comment]
@luke-jr Mind to solve the merge conflict?

Also, assigned 0.14 milestone, I think this is relatively trivial/uncontroversial refactoring.

**Comment by luke-jr** [comment]
Done

**Comment by MarcoFalke** [comment]
utACK 5394b39

File: src/wallet/walletdb.cpp [hunk 2]
```diff
@@ -810,6 +810,7 @@ void ThreadFlushWalletDB(const string& strFile)
                 if (nRefCount == 0)
                 {
                     boost::this_thread::interruption_point();
+                    const std::string& strFile = pwalletMain->strWalletFile;
```
Comment (jonasschnelli) [inline src/wallet/walletdb.cpp]: In theory, this should only be done when holding cs_wallet. I guess changing CWallet::strWalletFile to a `const std::string` is not possible with the current concept of setting the filename in a instance method.
But maybe I'm wrong.

File: src/wallet/wallet.h [hunk 3]
```diff
@@ -914,6 +917,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
     static std::string GetWalletHelpString(bool showDebug);
 
     /* Initializes the wallet, returns a new CWallet instance or a null pointer in case of an error */
```
Comment (jonasschnelli) [inline src/wallet/wallet.h]: Heh. This (unchanged) comment was wrong before and correct after this PR. :)

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -914,6 +917,7 @@ class CWallet : public CCryptoKeyStore, public CValidationInterface
...
+    static CWallet* CreateWalletFromFile(const std::string walletFile);
```
Comment (jonasschnelli) [inline src/wallet/wallet.h]: Maybe declare `CreateWalletFromFile()` private to avoid layer confusion?

File: src/wallet/walletdb.cpp (hunk 2 above)
Comment (luke-jr) [inline src/wallet/walletdb.cpp]: wallet.h says "This lock protects all the fields added by CWallet except for: ... strWalletFile (immutable after instantiation)"

File: src/wallet/wallet.h (hunk 4 above)
Comment (luke-jr) [inline src/wallet/wallet.h]: It's not really intended to be private...

**Comment by MarcoFalke** [comment]
@jonasschnelli Are your comments addressed by the feedback?

**Comment by jonasschnelli** [comment]
@jonasschnelli Are your comments addressed by the feedback?

Yes.
re-utACK 5394b3940dec1fd35952d344e6373fb7115c5490

**Comment by NicolasDorier** [comment]
Code Review ACK.

**Comment by gmaxwell** [comment]
utACK

**Comment by sipa** [comment]
utACK 5394b3940dec1fd35952d344e6373fb7115c5490

**Pull request merged**